File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
# 만들 수 있는 소수 개수 리턴
from itertools import permutations
from math import sqrt


def solution(numbers):
    answer = []

    # 1. 한자리수로 쪼개기
    nums = list(map(int, numbers))
    # 2. 소수 후보 숫자 만들기
    temp = []
    for i in range(1, len(numbers) + 1):  # numbers의 각 숫자들을 순열로 만들 경우
        temp += list(permutations(nums, i))

    per = []
    for t in temp:
        per.append("".join(map(str, t)))

    # 3. 소수 판별(소수: 1과 자기자신만 약수로 가짐)
    for p in per:
        p = int(p)

        # 0,1은 소수 아님
        if p < 2:
            continue

        # 소수이려면 제곱근 수까지 약수가 없어야.
        sqrt_num = int(sqrt(p))
        flag = True
        for i in range(2, sqrt_num + 1):
            if p % i == 0:  # 소수 아님
                flag = False
                break
        if flag:
            answer.append(p)
    print(answer)
    return len(set(answer))  # 중복 제거 후 개수 반환

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    HashSet<Integer> set = new HashSet<>();
    
    // 소수 체크
    public boolean isPrime(int num){
        // 1. 0과 1은 소수가 아님
        if(num == 0 || num ==1){
            return false;
        }
        // 2. 에라토스테네스의 체 limit 계산
        int limit = (int)Math.sqrt(num); // 루트 씌운 값까지 계산
        
        // 3. 에라토스 체에 따라 limit까지만 배수 여부 확인
        for(int i=2; i<limit + 1; i++){
            if(num % i == 0)
                return false;
        }
        return true;
    }
    
    public void makeNum(String comb, String others){
        // 1. 현재 조합을 set에 추가
        if(!comb.equals(""))
            set.add(Integer.valueOf(comb));
        
        for(int i=0; i<others.length(); i++){
            // 2.남은 숫자 들 중 하나를 더해 새로운 조합 만듦.
            makeNum(comb+others.charAt(i), others.substring(0,i) + others.substring(i+1));
        }
    }
    
    public int solution(String numbers) {
        int answer = 0;
        
        // 1. 숫자 조합 만들기 -> 재귀 사용
        makeNum("", numbers);
                
        // 2. 소수 개수 세기
        Iterator<Integer> iter = set.iterator();
        
        int cnt = 0;
        while(iter.hasNext()){
            int number = iter.next();
            
            if(isPrime(number))
                cnt ++;
        }
        
        // 3. 소수 개수 반환
        return cnt;
    }
}